language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/RequiredParameterNodeContext.java

```diff
@@ -0,0 +1,98 @@
+import io.ballerina.compiler.syntax.tree.NonTerminalNode;
+import io.ballerina.compiler.syntax.tree.QualifiedNameReferenceNode;
+import io.ballerina.compiler.syntax.tree.RequiredParameterNode;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.common.utils.completion.QNameReferenceUtil;
+import org.ballerinalang.langserver.commons.BallerinaCompletionContext;
+import org.ballerinalang.langserver.commons.completion.LSCompletionException;
+import org.ballerinalang.langserver.commons.completion.LSCompletionItem;
+import org.ballerinalang.langserver.completions.SnippetCompletionItem;
+import org.ballerinalang.langserver.completions.providers.AbstractCompletionProvider;
+import org.ballerinalang.langserver.completions.util.ContextTypeResolver;
+import org.ballerinalang.langserver.completions.util.Snippet;
+import org.ballerinalang.langserver.completions.util.SortingUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Handles the completions for the {@link io.ballerina.compiler.syntax.tree.RequiredParameterNode}.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.completion.spi.BallerinaCompletionProvider")
+public class RequiredParameterNodeContext extends AbstractCompletionProvider<RequiredParameterNode> {
+
+    public RequiredParameterNodeContext() {
+        super(RequiredParameterNode.class);
+    }
+
+    @Override
+    public List<LSCompletionItem> getCompletions(BallerinaCompletionContext context, RequiredParameterNode node)
+            throws LSCompletionException {
+
+        List<LSCompletionItem> completionItems = new ArrayList<>();
+        NonTerminalNode nodeAtCursor = context.getNodeAtCursor();
+        if (this.onQualifiedNameIdentifier(context, nodeAtCursor)) {
+            /*
+                Covers the Following
+                (1) function(mod:<cursor>)
+                (2) function(mod:T<cursor>)
+            */
+            QualifiedNameReferenceNode qNameRef = (QualifiedNameReferenceNode) nodeAtCursor;
+            completionItems.addAll(this.getCompletionItemList(QNameReferenceUtil
+                    .getTypesInModule(context, qNameRef), context));
+        } else {
+            ContextTypeResolver resolver = new ContextTypeResolver(context);
+            Optional<TypeSymbol> typeSymbol = node.apply(resolver);
+            /*
+              covers the following
+              function (int,int,..) re<cursor>
+            */
+            if (typeSymbol.isPresent() && typeSymbol.get().typeKind() == TypeDescKind.FUNCTION) {
+                Optional<TypeSymbol> returnType = ((FunctionTypeSymbol) typeSymbol.get()).returnTypeDescriptor();
+                if (returnType.isPresent() && returnType.get().typeKind() == TypeDescKind.NIL) {
+                    completionItems.add(new SnippetCompletionItem(context, Snippet.KW_RETURNS.get()));
+                }
+            } else {
```

Comment:
We are taking this decision based on the semantic information. The best approach would be to make the decision based on the syntax nodes.

Reply:
Done